### PR TITLE
 feat (engine): new vcs µService

### PR DESCRIPTION
**Pull request by fsamin**
TODO:
* Move Stash Client from API to µService
* Move Gitlab Client from API to µService
* Remove All client from API and call µService from API
* Improve Stash client code with cache
* Link Hooks/Poller to this µService
* Code Webhook with Github

### CODE REVIEW

File: engine/types.go [hunk 1]
```diff
@@ -36,7 +37,8 @@ type Configuration struct {
 		Swarm     swarm.HatcheryConfiguration     `toml:"swarm" comment:"Hatchery Swarm. Doc: https://ovh.github.io/cds/advanced/advanced.hatcheries.swarm/"`
 		VSphere   vsphere.HatcheryConfiguration   `toml:"vsphere" comment:"Hatchery VShpere. Doc: https://ovh.github.io/cds/advanced/advanced.hatcheries.vsphere/"`
 	} `toml:"hatchery"`
-	Hooks hooks.Configuration `toml:"hooks"`
+	Hooks hooks.Configuration `toml:"vcs" comment:"######################\n CDS Hooks Settings \n######################"`
```
Comment (sguiheux) [inline engine/types.go]: toml: hooks

File: engine/vcs/github/github.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package github
+
+import "github.com/ovh/cds/engine/api/cache"
```
Comment (sguiheux) [inline engine/vcs/github/github.go]: import (
)

File: engine/vcs/github/oauth_consumer.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net/url"
+	"strings"
+
+	"github.com/ovh/cds/sdk"
+
```
Comment (sguiheux) [inline engine/vcs/github/oauth_consumer.go]: no empty line

File: engine/vcs/vcs_handlers.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
+package vcs
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/ovh/cds/sdk"
```
Comment (sguiheux) [inline engine/vcs/vcs_handlers.go]: wrong place

File: engine/vcs/github/oauth_consumer.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	// with parameters : client_id, redirect_uri, scope, state
+	requestToken, err := generateHash()
+	if err != nil {
+		return "", "", err
+	}
+
+	val := url.Values{}
+	val.Add("client_id", g.ClientID)
+	//Leave the default value set in github
+	//val.Add("redirect_uri", g.AuthorizationCallbackURL)
```
Comment (bnjjj) [inline engine/vcs/github/oauth_consumer.go]: comment

File: engine/vcs/vcs.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+	log.Info("VCS> Starting service %s...", s.Cfg.Name)
+
+	//Instanciate a cds client
+	s.cds = cdsclient.NewService(s.Cfg.API.HTTP.URL)
+
+	//First register(heartbeat)
+	if err := s.doHeartbeat(); err != nil {
+		log.Error("VCS> Unable to register: %v", err)
+		return err
```
Comment (bnjjj) [inline engine/vcs/vcs.go]: 🌯

**Pull request merged**